### ISSUE TITLE
Implement attribute forwarding

### ISSUE THREAD

**Issue by tkarabela** [tkarabela/MfxVTK#3]
- [ ] input geometry attributes
- [x] output geometry attributes - implemented in https://github.com/tkarabela/MfxVTK/commit/c34154949d3770a5911d57325c407a471ce247d2
- [ ] extra attributes (UVs, etc.) - partial support in https://github.com/tkarabela/MfxVTK/commit/f0b60a92e7e9569c92e0855623462a26890b7a1d (copying instead of forwarding; only `uv{0..3}`, `color{0..3}` as supported by Blender host)

**Comment by eliemichel** [comment]
Handling at least UVs would be great, in particular I'd love to use the decimate effect for lowpoly backing but it needs to conserve UVs :)

**Comment by tkarabela** [comment]
I have added vertex color and UV handling in https://github.com/tkarabela/MfxVTK/commit/f0b60a92e7e9569c92e0855623462a26890b7a1d (`feature-mesh-optimization-attributes` branch). For UVs in particular, it doesn't work well out of the box; VTK uses point attributes, not vertex attributes, which messes up UV seams (see below). There are about five different decimation filters in VTK, but this issue is more fundamental - it looks like the topology has to be cut along UV seams, decimated, then sewn back together (ugh).

![Screenshot from 2020-11-02 23-29-16](https://user-images.githubusercontent.com/8166544/97926627-bd4f7100-1d63-11eb-8ce7-24dd073bc6f2.png)

![Screenshot from 2020-11-02 23-31-51](https://user-images.githubusercontent.com/8166544/97926642-c3455200-1d63-11eb-8b66-319f1ef90aae.png)

![Screenshot from 2020-11-02 23-32-01](https://user-images.githubusercontent.com/8166544/97926661-c80a0600-1d63-11eb-9da8-c7a63fdcdcf2.png)

**Comment by eliemichel** [comment]
Indeed, assigning UVs to points instead of vertices is an issue at UV island boundary. Splitting the mesh at these boundaries 1. could be an mfx modifier in itself and 2. wouldn't it change the result of the decimation? I noticed that the algorithm can be asked to conserve boundaries, but this means the point density won't change there? Or how to ensure that the boundaries get decimated the same way on connected islands? Is it that VTK does not have any notion of vertex/face corner attributes at all?

**Comment by tkarabela** [comment]
Yes, VTK has only two attribute attachments:

- "point" data, attached to 3D points in space
- "cell" data, attached to topology primitives of highest dimension (eg. all of `vtkVertex`, `vtkLine`, `vtkQuad`, `vtkHexahedron` are valid cells, but if you have a `vtkHexahedron`, you get just one value - there are functions to decompose it into quads/edges etc., but it's all implicit)

It kind of makes sense if you think about attributes as measurements of some quantity in 3D space, but for graphics stuff it's inconvenient.

I haven't yet looked into how the decimation behaves with pre-split mesh; turning off boundary decimation would indeed hinder the decimation process and letting boundaries be decimated would probably make them non-conformal (they would move and would have to be welded together with some tolerance).
